qiskit/quantum_info/operators/symplectic/pauli.py

```diff
@@ -0,0 +1,1006 @@
+
+    For example
+
+    .. code:
+
+        p = Pauli('-iXYZ')
+
+        print('P[0] =', repr(P[0]))
+        print('P[1] =', repr(P[1]))
+        print('P[2] =', repr(P[2]))
+        print('P[:] =', repr(P[:]))
+        print('P[::-1] =, repr(P[::-1]))
+    """
+
+    # pylint: disable = missing-param-doc, missing-type-doc
+    def __init__(self, z=None, x=None, phase=None, *, label=None):
+        """Initialize the Pauli.
+
+        When using the symplectic array input data both z and x arguments must
+        be provided, however the first (z) argument can be used alone for string
+        label, Pauli operator, or ScalarOp input data.
+
+        Args:
+            z (array or str or ScalarOp or Pauli): input data or symplectic z vector.
+            x (array): Optional, symplectic x vector.
+            phase (int or None): Optional, phase exponent from Z_4.
+            label (str): DEPRECATED, string label.
+
+        Raises:
+            QiskitError: if input array is invalid shape.
+        """
+        if isinstance(z, BasePauli):
+            # BasePauli initialization
+            base_z, base_x, base_phase = z._z, z._x, z._phase
+            if base_z.shape[0] != 1:
+                raise QiskitError("Input is not a single Pauli")
+        elif x is not None:
+            base_z, base_x, base_phase = self._from_array(z, x, phase)
+            if base_z.shape[0] != 1:
+                raise QiskitError("Input is not a single Pauli")
+        elif isinstance(z, str):
+            base_z, base_x, base_phase = self._from_label(z)
+        elif label is not None:  # DEPRECATED
+            base_z, base_x, base_phase = self._from_label_deprecated(label)
+        elif isinstance(z, ScalarOp):
+            base_z, base_x, base_phase = self._from_scalar_op(z)
+        elif isinstance(z, (QuantumCircuit, Instruction)):
+            base_z, base_x, base_phase = self._from_circuit(z)
+        else:
+            raise QiskitError("Invalid input data for Pauli.")
+
+        # Add phase
+        if phase is not None:
+            base_phase = np.mod(
+                np.sum(np.logical_and(base_z, base_x),
+                       axis=1, dtype=np.int) + phase, 4)
+
+        # Initialize BasePauli
+        super().__init__(base_z, base_x, base_phase)
+
+    def __repr__(self):
+        """Display representation."""
+        return "Pauli('{}')".format(self.__str__())
+
+    def __str__(self):
+        """Print representation."""
+        if self.num_qubits > 100:
+            front = self[-2:].to_label()
+            back = self[:2].to_label()
+            return "%s...%s" % (front, back)
+        return self.to_label()
+
+    def __eq__(self, other):
+        """Test if two Paulis are equal."""
+        if not isinstance(other, Pauli):
+            return False
+        return (np.all(np.mod(self._phase, 4) == np.mod(other._phase, 4))
+                and np.all(self._z == other._z)
+                and np.all(self._x == other._x))
```

Comment:
It's good to compare the shape first. Numpy shows the following message with paulis with different sizes.
`DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  and np.all(self._z == other._z)`
```suggestion
        if self._z.shape != other._z.shape or self._x.shape != other._x.shape:
            return False 
        return (np.all(np.mod(self._phase, 4) == np.mod(other._phase, 4))
                and np.all(self._z == other._z)
                and np.all(self._x == other._x))
```